### PR TITLE
Add mockPort utility for typed interface mocking

### PR DESCRIPTION
**Pull request by cteyton**
## Explanation

Introduces `mockPort<T>()`, a new test utility for creating fully-typed mocks of interfaces (ports) that have no class to walk at runtime. This complements the existing `createMockInstance` which works with classes.

The utility:
- Lazily backs every member with a `jest.fn()`, ensuring the mock is complete by construction
- Provides type-safe stub configuration that's checked against the port contract
- Avoids the pitfalls of `as unknown as jest.Mocked<T>` casts, which disable structural type checking
- Handles edge cases like thenable detection and special runtime properties that must never be mocked

Refactors 15+ test files across the git package to use `mockPort` instead of manual object literals with type casts, significantly reducing boilerplate and improving type safety.

## Type of Change

- [x] New feature
- [x] Refactoring
- [x] Documentation

## Affected Components

- **Domain packages affected**: `@packmind/test-utils`
- **Frontend / Backend / Both**: Both (test utilities)
- **Breaking changes**: None

## Testing

- [x] Unit tests added/updated
  - Added comprehensive test suite (`mockPort.spec.ts`) covering:
    - Lazy mock creation and identity preservation
    - Call recording through mocks
    - Stub implementations (both real functions and pre-built mocks)
    - Post-construction configuration
    - Data member handling
    - Thenable detection (ensuring mocks don't behave as thenables when awaited)
- [x] Integration tests via refactored specs
  - 15+ existing test files refactored to use `mockPort`, validating the utility works correctly in real test scenarios
  - All existing tests continue to pass with the new utility

**Test Details:**
The new `mockPort.spec.ts` validates all core behaviors: lazy initialization, mock identity, call tracking, stub wrapping, and special property handling. The refactored test files demonstrate the utility's effectiveness in reducing boilerplate while maintaining full type safety and test functionality.

## TODO List

- [x] CHANGELOG Updated (via documentation in CLAUDE.md)
- [x] Documentation Updated (added section to CLAUDE.md with usage examples)

## Reviewer Notes

The implementation uses a Proxy to lazily create mocks and carefully avoids mocking properties that the JavaScript runtime probes (like `then`, `toJSON`, `$$typeof`) which would break test behavior. The type system ensures data members must be provided while methods are optional, catching contract violations at build time rather than runtime.

https://claude.ai/code/session_01BXLu2mQf7NhTSS2q4TS4Lr

### CODE REVIEW

File: packages/test-utils/src/mockPort.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ * A member that is *not* a method has to be given: nothing sensible can be
+ * conjured for it, since the proxy cannot tell a data member from a method at
+ * runtime and would hand out a `jest.fn()` where the port declares a value.
+ */
+export type PortStubs<T> = Partial<{
+  [K in MethodKeys<T>]: T[K] extends (...args: infer A) => infer R
+    ? T[K] | jest.Mock<R, A>
+    : never;
+}> & { [K in DataKeys<T>]: T[K] };
```
[the remark this answers is not in the document]
Comment (cteyton) [reply packages/test-utils/src/mockPort.ts]: Confirmed and fixed in 445e7a4. The stub type now reads `NonNullable<T[K]>`, as suggested, so an optional method such as `close?(): void` can be given an implementation instead of collapsing to `never`. Two tests pin it: one seeds `close` and asserts the call is recorded, one leaves it out and asserts it is still mocked.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * supplied, and one whose shape is mostly data is better mocked by hand.
+ */
+export function mockPort<T extends object>(
+  ...args: PortStubsArgs<T>
+): jest.Mocked<T> {
+  const [stubs = {} as PortStubs<T>] = args;
+  const members = new Map<string, unknown>();
+
+  const isMockable = (member: string | symbol): member is string =>
+    typeof member === 'string' && !NEVER_MOCKED.has(member);
```
[the remark this answers is not in the document]
Comment (cteyton) [reply packages/test-utils/src/mockPort.ts]: Confirmed and fixed in 445e7a4, though not by auto-mocking symbols: the runtime probes `Symbol.iterator`, `Symbol.toPrimitive` and friends on any object it is handed, so answering them with a `jest.fn()` would make the mock claim behaviour the port never declared.

Instead the type no longer promises what the proxy cannot deliver. Only plainly named methods are auto-mocked; symbol-keyed members — and members named after the probes in `NEVER_MOCKED`, which had the same hole — now sit in the "has to be given" bucket alongside data members, so `mockPort<IPortWithSymbolMethod>()` fails to compile rather than handing back `undefined`. Seeding one works: `Reflect.ownKeys` replaces `Object.entries` so symbol-keyed stubs survive, and the proxy returns a given member before deciding whether it may mock. A `@ts-expect-error` test pins the rejection and another asserts the seeded symbol member records its calls.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+ * declares a value, and a symbol-keyed or probe-named method would come back
+ * `undefined`.
+ */
+export type PortStubs<T> = Partial<{
+  [K in AutoMockedKeys<T>]: NonNullable<T[K]> extends (
+    ...args: infer A
+  ) => infer R
+    ? NonNullable<T[K]> | jest.Mock<R, A>
+    : never;
+}> & { [K in GivenKeys<T>]: T[K] };
```
[the remark this answers is not in the document]
Comment (cteyton) [reply packages/test-utils/src/mockPort.ts]: Confirmed and fixed in 49835b4 — with one deliberate exception, because the finding holds for two kinds of member that the proxy treats differently.

The cause was the mapped type: `{ [K in Exclude<keyof T, …>]: T[K] }` is not homomorphic, so it dropped the optional modifiers. `Pick` keeps them, and excluded **methods** — symbol-keyed or probe-named — now follow the port's own optionality: `mockPort<{ [tag]?(): void; toJSON?(): unknown }>()` compiles, and the `undefined` the proxy answers with is exactly what such a port allows.

A **data** member stays required even when the port marks it optional. The proxy has only the member's name to go on, so omitting `name?: string` does not yield `undefined` — it yields a `jest.fn()`, a function standing where a value belongs, which is the lie this helper exists to prevent. I found that by writing the test for the fix as reported and watching it fail on `expect(port.name).toBeUndefined()`; the case is now pinned by a test and a `@ts-expect-error`.

Worth noting the finding understated its own scope: optional data members were caught by the same bug, not just symbol-keyed and reserved-name ones.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.spec.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+describe('mockPort', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('backs every member with a jest mock', () => {
+    const port = mockPort<ITestPort>();
+
+    expect(jest.isMockFunction(port.findById)).toBe(true);
```
[the remark this answers is not in the document]
Comment (cteyton) [reply packages/test-utils/src/mockPort.spec.ts]: Agreed and fixed in c5f4c70. All three `jest.isMockFunction(...)` assertions are gone:

- the data-member case now calls `port.findById('id')` and asserts the call was recorded
- the optional-method case calls `port.close?.()` and asserts the same
- the first test was dropped outright rather than reworded — `records the calls made through the port` already covers that an unstubbed member is mocked, so restating it as behaviour would only have duplicated that test

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    ? [stubs?: PortStubs<T>, options?: MockPortOptions]
+    : [stubs: PortStubs<T>, options?: MockPortOptions];
+
+type UnknownFunction = (...args: unknown[]) => unknown;
+
+function refuseUnstubbedCall(member: string): UnknownFunction {
+  return () => {
+    throw new Error(
+      `mockPort: '${member}' was called but was never stubbed. Stub it, or drop the strict option if answering undefined is what this test wants.`,
+    );
```
[the remark this answers is not in the document]
Comment (cteyton) [reply packages/test-utils/src/mockPort.ts]: Agreed and fixed in f97e2ed. I checked the directive rather than taking it on trust: the standard's `paths` glob is `**/packages/**/*.ts`, so it covers this helper even though it only ever runs in tests.

`UnstubbedPortCallError` now carries the member it refused, which turned out to pay for itself in the spec — the two tests asserted on a message regex (`/was called but was never stubbed/`, `/findById/`) and now assert on the type and the named member instead.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+ * are typed against the port, so a fixture that drifts from the contract fails
+ * the build:
+ *
+ * ```ts
+ * const gitRepo = mockPort<IGitRepo>();
+ * gitRepo.getFileOnRepo.mockResolvedValue({ sha: 'sha', content: 'content' });
+ *
+ * // or seed implementations up front
+ * const accounts = mockPort<IAccountsPort>({
+ *   getUserById: async () => user,
```
Comment (vincent-psarga) [inline packages/test-utils/src/mockPort.ts]: I'm a bit dubious about this example. Shouldn't it be:

```ts
const accounts = mockPort<IAccountsPort>({
   getUserById: jest.fn().resolvesWith(user)
});
```

Otherwise, we may not be sure that the mock is really a mock.

File: packages/git/src/infra/repositories/github/GithubRepository.spec.ts [hunk 7]
```diff
@@ -31,12 +31,14 @@ describe('GithubRepository', () => {
...
+    // rather than with mockPort: only what this suite drives is stubbed.
     mockAxiosInstance = {
       interceptors: {
-        request: { use: jest.fn() },
-        response: { use: jest.fn() },
+        request: { use: jest.fn(), eject: jest.fn(), clear: jest.fn() },
+        response: { use: jest.fn(), eject: jest.fn(), clear: jest.fn() },
       },
-    } as unknown as jest.Mocked<typeof axios>;
+    } as Partial<jest.Mocked<typeof axios>> as jest.Mocked<typeof axios>;
```
Comment (vincent-psarga) [inline packages/git/src/infra/repositories/github/GithubRepository.spec.ts:41]: is the as Partial<...> needed here?

File: packages/git/src/application/services/ResolvedGitRepoService.spec.ts [hunk 8]
```diff
@@ -9,16 +9,10 @@ import {
...
 import { instrumentComponents } from '@packmind/node-utils';
 
-const stubGitRepoInstance = (): jest.Mocked<IGitRepo> =>
-  ({
-    getFileOnRepo: jest.fn(),
-    commitFiles: jest.fn(),
-    listDirectoriesOnRepo: jest.fn(),
-    checkDirectoryExists: jest.fn(),
-  }) as unknown as jest.Mocked<IGitRepo>;
+const stubGitRepoInstance = (): jest.Mocked<IGitRepo> => mockPort<IGitRepo>();
```
Comment (vincent-psarga) [inline packages/git/src/application/services/ResolvedGitRepoService.spec.ts]: It's a bit weird to use `mockPort` to mock something that is not a port.

Shouldn't we use `mockInterface` so this is clearer?

File: packages/test-utils/src/invalidInput.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ * });
+ * ```
+ *
+ * The cast such a test needs lives here, once and under a name, rather than as
+ * an `as unknown as` in each spec. It buys no type safety - nothing can, since
+ * the whole point is a value the type forbids - but it tells a reader which
+ * casts are deliberate, so the ones that are merely a mock that stopped
+ * matching its port stand out instead of blending in.
+ */
+export function invalidInput<T>(value: unknown): T {
```
Comment (vincent-psarga) [inline packages/test-utils/src/invalidInput.ts:20]: I like that :)

**Review by vincent-psarga — CHANGES_REQUESTED** [review]

File: packages/git/src/application/services/ResolvedGitRepoService.spec.ts (hunk 8 above)
Comment (cteyton) [reply packages/git/src/application/services/ResolvedGitRepoService.spec.ts]: Agreed — renamed throughout in 9c4888c. You're right that the name claimed more than the helper does: half of what it mocks here is a `GitProviderService`, a `Cache`, a use case reached through its type. It never cared what the type was, so `mockInterface` is the honest name.

`PortStubs` → `InterfaceStubs` and `MockPortOptions` → `MockInterfaceOptions` followed, and the strict-mode error drops the word entirely as `UnstubbedCallError`. The doc comment and `test-utils/CLAUDE.md` now say "a port, a service reached through its type, anything structural" rather than implying ports only.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/git/src/infra/repositories/github/GithubRepository.spec.ts (hunk 7 above)
Comment (cteyton) [reply packages/git/src/infra/repositories/github/GithubRepository.spec.ts:41]: Yes — I tried removing it to be sure. With a single `as jest.Mocked<typeof axios>`, `tsc` rejects the literal:

```
error TS2352: Conversion of type '{ interceptors: { request: { use: jest.Mock<…> … } } }'
to type 'Mocked<AxiosStatic>' may be a mistake because neither type sufficiently
overlaps with the other.
```

`AxiosStatic` is callable and carries `defaults`, `getUri`, `create`, the verb methods… so a literal holding only `interceptors` doesn't overlap enough for a direct assertion. `Partial<jest.Mocked<typeof axios>>` is a type the literal *does* satisfy, and the second step widens it back — which is the same trick as `as unknown as`, except it still checks the members that are present. That's why `interceptors` had to grow `eject` and `clear`: `AxiosInterceptorManager` declares them, and this cast notices, where `as unknown as` did not.

An axios instance is the one shape in this package `mockInterface` can't take — it's callable and mostly data, so the proxy would answer `defaults` with a `jest.fn()`. Hence the hand-written literal here and in the two GitLab specs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/test-utils/src/mockPort.ts (hunk 6 above)
Comment (cteyton) [reply packages/test-utils/src/mockPort.ts]: It is really a mock — that was the doc's fault for not saying so. A plain implementation is wrapped in a `jest.fn()` on the way in, so the member records its calls exactly like a hand-built mock, and a test pins it (`when a member is stubbed with an implementation › still records the calls`).

Where the two forms differ is checking, and it favours the implementation: `async () => user` is compared against `IAccountsPort['getUserById']`, while `jest.fn()` is typed `any`, so `jest.fn().mockResolvedValue(somethingWrong)` sails through — which is the hole this PR set out to close. Both forms are accepted, since a prebuilt mock is sometimes what you have.

I've rewritten that part of the doc comment in 9c4888c to say the implementation is wrapped, to recommend it for that reason, and the example now shows the assertion working:

```ts
const accounts = mockInterface<IAccountsPort>({
  getUserById: async () => user, // checked against IAccountsPort
});

expect(accounts.getUserById).toHaveBeenCalledWith(user.id); // still a mock
```

(Small aside: `jest.fn().resolvesWith(...)` isn't a jest API — the equivalent would be `mockResolvedValue`.)

---
_Generated by [Claude Code](https://claude.ai/code)_
